### PR TITLE
Bond Source Prototype

### PR DESCRIPTION
**Pull request by jmcragun**
This is my attempt at implementing `IBTK::CartGridFunction` for bond source terms. I interpreted this as computing alpha as given in the new model (equations 8-10) it is heavily inspired by `ActivatedPlateletSource` and since they use the same cell variables (as far as I can tell, save for a few constants) many functions like `setDataOnPatchHierarchy` are exactly the same. So please let me know if those need to be updated in some nontrivial way.

I have not built this yet so do not merge until I can confirm that it is free of compilations (I'm still not totally sure how to build the project just yet)

### CODE REVIEW

**Comment by jmcragun** [comment]
Actually, it just occurred to me that I need to include these files in the makefile. Will do that in a sec.

File: BondSource.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+#include <SAMRAI_config.h>
+#include <math.h>
+// local include
+#include "BondSource.h"
+
+/////////////////////////////// PUBLIC ///////////////////////////////////////
+
+BondSource::BondSource(SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> phi_u_var,
+                    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> phi_a_var,
+                    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> w_var,
```
Comment (abarret) [inline BondSource.cpp:28]: Take a look at the modifications I made to the platelet sources for the wall sites. Wall sites are not controlled by the advection diffusion integrator, we have to get them in a different way.
https://github.com/abarret/clotting_model/blob/0251592732b153423aab269b6a2d2159b2d282f1/PlateletSource.h#L89

File: BondSource.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+} // setDataOnPatchHierarchy
+
+void BondSource::setDataOnPatch(const int data_idx,
+                                        Pointer<Variable<NDIM>> /*var*/,
+                                        Pointer<Patch<NDIM>> patch,
+                                        const double /*data_time*/,
+                                        const bool initial_time,
+                                        Pointer<PatchLevel<NDIM>> /*patch_level*/)
+{
+    // Calling this "Alpha data" because I believe the goal for this method is to compute alpha (10)
```
Comment (abarret) [inline BondSource.cpp:141]: This should be computing alpha - beta * z

**Review by abarret — CHANGES_REQUESTED** [review]
You should also add BondSource.cpp to the CMakeLists.txt file

File: BondSource.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+#include <SAMRAI_config.h>
+#include <math.h>
+// local include
+#include "BondSource.h"
+
+/////////////////////////////// PUBLIC ///////////////////////////////////////
+
+BondSource::BondSource(SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> phi_u_var,
+                    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> phi_a_var,
+                    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> w_var,
```
Comment (abarret) [inline BondSource.cpp:28]: We need some way to set `z_var` and `d_sigma_var`. I would suggest setting them in the constructor (in place of `w_var` which is no longer needed).

File: BondSource.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        // grab the index
+        const CellIndex<NDIM>& idx = ci();
+        // grab the var values w/ the index
+        double phi_u = (*phi_u_data)(idx);
+        double phi_u = (*phi_u_data)(idx);
+        double w = (*w_data)(idx);
+        double z = (*z_data)(idx);
+        // Compute alpha data
+        double alpha = (d_a0 * phi_a * phi_a) + (d_a0w * (d_w_mx - w) * phi_u);
+        double beta = 0.0;
```
Comment (abarret) [inline BondSource.cpp:169]: Beta is defined in between the dimension checks.
```suggestion
```

**Review by abarret — CHANGES_REQUESTED** [review]

File: BondSource.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    Pointer<CellData<NDIM, double>> w_data = patch->getPatchData(d_w_idx);
+    const Box<NDIM>& patch_box = patch->getBox();
+    // begin cell loop
+    for (CellIterator<NDIM> ci(patch_box); ci; ci++)
+    {
+        // grab the index
+        const CellIndex<NDIM>& idx = ci();
+        // grab the var values w/ the index
+        double phi_u = (*phi_u_data)(idx);
+        double phi_u = (*phi_u_data)(idx);
```
Comment (abarret) [inline BondSource.cpp:166]:
```suggestion
        double phi_a = (*phi_a_data)(idx);
```

File: BondSource.cpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            const double trace = (*sig_data)(idx, 0) + (*sig_data)(idx, 1);
+            const double y_brackets = std::sqrt(trace / (z + 1.0e-12));
+            double beta = d_beta_fcn(y_brackets);
+        #endif
+        #if (NDIM == 3)
+            const double trace = (*sig_data)(idx, 0) + (*sig_data)(idx, 1) + (*sig_data)(idx, 2);
+            const double y_brackets = std::sqrt(trace / (z + 1.0e-12));
+            double beta = d_beta_fcn(y_brackets);
+        #endif
+        (*bond_data)(idx) = alpha - beta * z
```
Comment (abarret) [inline BondSource.cpp:181]:
```suggestion
        (*bond_data)(idx) = alpha - beta * z;
```

File: BondSource.h [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+        // Beta function pointer
+        std::function<double(double)> d_beta_fcn;
+
+        // OTHER THINGS
+        BondSource(const BondSource& from) = delete;
+        BondSource& operator=(const BondSource& that) = delete;
+        SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM>> d_phi_u_var, d_phi_a_var, d_z_var, d_sig_var;
+        SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> d_adv_diff_hier_integrator;
+    }
```
Comment (abarret) [inline BondSource.h:112]:
```suggestion
    };
```

**Review by abarret — CHANGES_REQUESTED** [review]

**Comment by abarret** [comment]
I fixed a few compilation things. I think it looks good otherwise. I'll go ahead and merge so we can finish wrapping things up.

**Comment by jmcragun** [comment]
Thanks! Merging now

**Comment by jmcragun** [comment]
Thanks!

**Pull request merged**